sdks/java/core/src/main/java/org/apache/beam/sdk/io/WriteFiles.java

```diff
@@ -122,27 +126,35 @@
    * Creates a {@link WriteFiles} transform that writes to the given {@link FileBasedSink}, letting
    * the runner control how many different shards are produced.
    */
-  public static <T> WriteFiles<T> to(FileBasedSink<T> sink) {
+  public static <InT, DestT, OutT> WriteFiles<InT, DestT, OutT> to(
```

Comment:
Is InT same as UserT? I think it's important to use consistent terminology for type parameters. There's 3 roles - the PCollection element type, the raw type being written to a FileBasedSink, and the destination type. I'd suggest to uniformly call them throughout the PR: InputT (also fine with UserT as long as it's uniform), OutputT, DestinationT.